### PR TITLE
[PowerPC] Support -fpatchable-function-entry

### LINKED ISSUE #57031: [clang][PPC] support for -mprofile-kernel
From an LKML thread, it seems like the Linux kernel makes use of `-mprofile-kernel`.
https://lore.kernel.org/llvm/20220809095907.418764-1-naveen.n.rao@linux.vnet.ibm.com/
cc @nemanjai @nemanja-ibm

### LINKED ISSUE #63220: [PowerPC] call to _mcount emitted by -pg needs to be emitted BEFORE function prolog
From an LKML thread: https://lore.kernel.org/llvm/20230609034501.407971-1-naveen@kernel.org/

```
Ftrace on ppc32 expects a three instruction sequence at the beginning of
each function when specifying -pg:
	mflr	r0
	stw	r0,4(r1)
	bl	_mcount

This is the case with all supported versions of gcc. Clang however emits
a branch to _mcount after the function prologue
```
Example: https://godbolt.org/z/8r83PKorK

As a result, support for function tracing is being disabled in the linux kernel for ppc32 when built with clang.

cc @nemanjai @nemanja-ibm

### PR DESCRIPTION
**Pull request by chenzheng1030**
For now only PPC big endian Linux 32 and 64 bit are supported.

PPC little endian Linux has XRAY support for 64-bit.
PPC AIX has different patchable function entry implementations.

Fixes #63220
Fixes #57031

### CODE REVIEW

File: clang/test/Sema/patchable-function-entry-attr-aix.cpp [hunk 1]
```diff
@@ -0,0 +1,4 @@
+// RUN: %clang_cc1 -triple powerpc64-ibm-aix-xcoff -fsyntax-only -verify %s
+
+// expected-error@+1 {{'patchable_function_entry' attribute is not yet supported on AIX}}
+__attribute__((patchable_function_entry(0))) void f();
```
[the remark this answers is not in the document]
Comment (chenzheng1030) [reply clang/test/Sema/patchable-function-entry-attr-aix.cpp:4]: I don't see a clean way to do that. The error message from in `clang/test/Sema/patchable-function-entry-attr.cpp` is to detect the missing targets in `clang/include/clang/Basic/Attr.td`. But this patch declares ppc/ppc64 as supported for Linux. So we have to emit the error message in some other place for example, in `clang/lib/Sema/SemaDeclAttr.cpp`.

**Comment by maryammo** [comment]
In `PPCTargetMachine::PPCABI computeTargetABI` it specifies the ABI  with this 
```
switch (TT.getArch()) {
 case Triple::ppc64le:
  return PPCTargetMachine::PPC_ABI_ELFv2;
 case Triple::ppc64:
  if (TT.isPPC64ELFv2ABI())
   return PPCTargetMachine::PPC_ABI_ELFv2;
  else
   return PPCTargetMachine::PPC_ABI_ELFv1;
 default:
  return PPCTargetMachine::PPC_ABI_UNKNOWN;
 }
```
`isPPC64ELFv2ABI()`  is defined as 
```
/// Tests whether the target 64-bit PowerPC big endian ABI is ELFv2.
 bool isPPC64ELFv2ABI() const {
  return (getArch() == Triple::ppc64 &&
      ((getOS() == Triple::FreeBSD &&
       (getOSMajorVersion() >= 13 || getOSVersion().empty())) ||
       getOS() == Triple::OpenBSD || isMusl()));
 }
```
So PPC64 uses ELFv2 for Triple::OpenBSD. We probably want to  diagnose this OS for PC64, since with ELFv2 we might emit separate local  and global entry points which means only certain values can be passed to -fpatchable-function-entry option.

File: clang/test/Sema/patchable-function-entry-attr-aix.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (MaskRay) [reply clang/test/Sema/patchable-function-entry-attr-aix.cpp:4]: Some test directory adopt `rm -rf %t && split-file %s %t && cd %t` (https://llvm.org/docs/TestingGuide.html#extra-files)

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 2]
```diff
@@ -6681,7 +6681,9 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
     StringRef S0 = A->getValue(), S = S0;
     unsigned Size, Offset = 0;
     if (!Triple.isAArch64() && !Triple.isLoongArch() && !Triple.isRISCV() &&
-        !Triple.isX86())
+        !Triple.isX86() &&
+        !(!Triple.isOSAIX() && (Triple.getArch() == llvm::Triple::ppc ||
```
Comment (MaskRay) [inline clang/lib/Driver/ToolChains/Clang.cpp:6685]: little-endian variants (ppcle ppc64le) are omitted here.

Use `isPPC`

File: clang/test/Driver/fpatchable-function-entry.c [hunk 3]
```diff
@@ -6,15 +6,20 @@
 // RUN: %clang -target loongarch64 %s -fpatchable-function-entry=1,0 -c -### 2>&1 | FileCheck %s
 // RUN: %clang -target riscv32 %s -fpatchable-function-entry=1,0 -c -### 2>&1 | FileCheck %s
 // RUN: %clang -target riscv64 %s -fpatchable-function-entry=1,0 -c -### 2>&1 | FileCheck %s
+// RUN: %clang -target powerpc-unknown-linux-gnu %s -fpatchable-function-entry=1,0 -c -### 2>&1 | FileCheck %s
```
Comment (MaskRay) [inline clang/test/Driver/fpatchable-function-entry.c:9]: Use `--target=` for new tests. `-target ` has been deprecated for a long time

File: llvm/test/CodeGen/PowerPC/patchable-function-entry.ll [hunk 4]
```diff
@@ -0,0 +1,49 @@
+; RUN: llc -mtriple=powerpc-unknown-linux-gnu %s -o - | FileCheck %s --check-prefixes=CHECK,PPC32
```
Comment (MaskRay) [inline llvm/test/CodeGen/PowerPC/patchable-function-entry.ll:1]: `-mtriple=powerpc`. Prefer a generic ELF triple to a Linux specific one.

Suggest 2 RUN lines, `powerpc` and `powerpc64le`, to catch both 32-bit/64-bit and LE/BE.

File: llvm/test/CodeGen/PowerPC/patchable-function-entry.ll [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+; PPC64:       .p2align    3, 0x0
+; PPC32-NEXT:  .long   .Lfunc_begin2
+; PPC64-NEXT:  .quad   .Lfunc_begin2
+  ret void
+}
+
+define void @f3() "patchable-function-entry"="1" "patchable-function-prefix"="2" {
+; CHECK-LABEL: .Ltmp0:
+; CHECK-COUNT-2: nop
+; CHECK-LABEL: f3:
```
Comment (MaskRay) [inline llvm/test/CodeGen/PowerPC/patchable-function-entry.ll:39]: Test addis/addi, instructions at the global entry point.

Per https://gcc.gnu.org/bugzilla/show_bug.cgi?id=99888#c5 `#c8` and `#15`, GCC will change the current strategy to match us :)

**Review by MaskRay — CHANGES_REQUESTED** [review]

**Comment by MaskRay** [comment]
> So PPC64 can use ELFv2 for `Triple::OpenBSD`. We probably want to diagnose this OS for PPC64, since with ELFv2 we might emit separate local and global entry points which means only certain values can be passed to `-fpatchable-function-entry` option.

This patch matches the intended end result: NOPs before the global entry and after the local entry.

The current way GCC handles `N,M` where M>0 is undesired and will hopefully be soon changed: https://gcc.gnu.org/pipermail/gcc-patches/2024-May/651025.html , matching what I suggested in 2022: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=99888#c5

File: llvm/test/CodeGen/PowerPC/patchable-function-entry.ll (hunk 5 above)
Comment (chenzheng1030) [reply llvm/test/CodeGen/PowerPC/patchable-function-entry.ll:41]: Updated.

**Review by chenzheng1030 — COMMENTED** [review]
> So PPC64 can use ELFv2 for Triple::OpenBSD. We probably want to diagnose this OS for PPC64, since with ELFv2 we might emit separate local and global entry points which means only certain values can be passed to -fpatchable-function-entry option.

For targets with ABI elfv2 which means there are dual entries, this patch can generate expected patchable entry:
For -mtriple=powerpc64-unknown-freebsd-unknown
```
.Ltmp0:
	nop
	nop
f3:                                     # @f3
.Lfunc_begin0:
	.cfi_startproc
.Lfunc_gep0:
	addis 2, 12, .TOC.-.Lfunc_gep0@ha
	addi 2, 2, .TOC.-.Lfunc_gep0@l
.Lfunc_lep0:
	.localentry	f3, .Lfunc_lep0-.Lfunc_gep0
# %bb.0:                                # %entry
	nop
```

**Review by MaskRay — APPROVED** [review]

**Pull request merged**